### PR TITLE
Setup devise

### PR DESCRIPTION
**Pull request by cloudkhoo96**

### CODE REVIEW

**Comment by cloudkhoo96** [comment]
TODO:
- [ ] Set up Devise
- [ ] User can sign up
- [ ] User can sign in
- [ ] User can sign out
- [ ] User can edit email, username, password
- [ ] Enable sign in with either email or username in the same field
- [ ] Articles Controller - Only logged in for all except index and show
- [ ] Comments controller - Only logged in for all

**Comment by cloudkhoo96** [comment]
TODO:
- [x] Set up Devise
- [x] User can sign up
- [x] User can sign in
- [x] User can sign out
- [x] User can edit email, username, password
- [ ] Enable sign in with either email or username in the same field
- [x ] Articles Controller - Only logged in for all except index and show
- [x] Comments controller - Only logged in for all

**Comment by cloudkhoo96** [comment]
TODO:
- [x] Set up Devise
- [x] User can sign up
- [x] User can sign in
- [x] User can sign out
- [x] User can edit email, username, password
- [x] Enable sign in with either email or username in the same field
- [x] Articles Controller - Only logged in for all except index and show
- [x] Comments controller - Only logged in for all

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  attr_writer :login
```
Comment (synmal) [inline app/models/user.rb:7]: This is not needed. Please refer the comment below

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  validate :validate_username
+
+  def login
+    @login || self.username || self.email
+  end
```
Comment (synmal) [inline app/models/user.rb:12]: There's no need for `@login`. You can just create a publicmethod which does the similar as above

```ruby
  def login
    self.username || self.email
  end
```

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -5,13 +5,24 @@
...
+      <ul>
+        <li><%= link_to "Sign Up", new_user_registration_path %></li>
+        <li><%= link_to "Sign In", new_user_session_path %></li>
+      </ul>
+    <% else %>
+    <ul>
+      <li><%= link_to "Edit Account", edit_user_registration_path %></li>
+      <li><%= link_to "Sign out", destroy_user_session_path, data: { turbo_method: :delete } %></li>
+    </ul>
+    <% end %>
```
Comment (synmal) [inline app/views/layouts/application.html.erb:23]: To simplify this
```erb
<ul>
  <% if !user_signed_in? %>
    <li><%= link_to "Sign Up", new_user_registration_path %></li>
    <li><%= link_to "Sign In", new_user_session_path %></li>
  <% else %>
    <li><%= link_to "Edit Account", edit_user_registration_path %></li>
    <li><%= link_to "Sign out", destroy_user_session_path, data: { turbo_method: :delete } %></li>
  <% end %>
</ul>
```

File: app/models/user.rb (hunk 1 above)
Comment (cloudkhoo96) [reply app/models/user.rb]: Done

File: app/models/user.rb (hunk 2 above)
Comment (cloudkhoo96) [reply app/models/user.rb]: Done

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (cloudkhoo96) [reply app/views/layouts/application.html.erb]: Done

**Comment by synmal** [comment]
LGTM

**Pull request merged**